core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,304 @@
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * XML Reader Batch Source Plugin
+ * It is used to read XML files from HDFS with specified file properties and filters.
+ * It parses the read file into specified Output Schema.
+ * A {@link FileBatchSource} to use any distributed file system as a Source.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name("XMLReader")
+@Description("Batch source for XML read from HDFS")
+public class XMLReaderBatchSource extends ReferenceBatchSource<LongWritable, Object, StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(XMLReaderBatchSource.class);
+
+  public static final Schema DEFAULT_XML_SCHEMA = Schema.recordOf(
+    "xmlSchema",
+    Schema.Field.of("offset", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("filename", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("record", Schema.of(Schema.Type.STRING))
+  );
+
+  private static final Gson GSON = new Gson();
+  private static final Type ARRAYLIST_PREPROCESSED_FILES  = new TypeToken<ArrayList<String>>() { }.getType();
+
+  private KeyValueTable processedFileTrackingTable;
+  private final XMLReaderConfig config;
+  private FileSystem fileSystem;
+  private Path tempDirectoryPath;
+
+  public XMLReaderBatchSource(XMLReaderConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  @VisibleForTesting
+  XMLReaderConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    config.validateConfig();
+    pipelineConfigurer.getStageConfigurer().setOutputSchema(DEFAULT_XML_SCHEMA);
+    pipelineConfigurer.createDataset(config.tableName, KeyValueTable.class.getName());
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_PATH_NAME, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, config.nodePath);
+    if (StringUtils.isNotEmpty(config.pattern)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PATTERN, config.pattern);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, config.actionAfterProcess);
+    if (StringUtils.isNotEmpty(config.targetFolder)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, config.targetFolder);
+    }
+
+    setFileTrackingInfo(context, conf);
+
+    //create a temporary directory, in which XMLRecordReader will add file tracking information.
+    fileSystem = FileSystem.get(conf);
+    long startTime = context.getLogicalStartTime();
+    //create temp file name using start time to make it unique.
+    String tempDirectory = "/tmp/" + config.tableName + startTime;
```

Comment:
There is no guarantee that ``/tmp`` will always exist on HDFS, and all users will have access to write to it. IMO, if you have to use a directory, users should be allowed to change it. However, since that influences the plugin configuration, @nitinmotgi please advice if it is ok to do so.